contracts/transfer/src/state.rs

```diff
@@ -223,28 +228,168 @@ impl TransferState {
 
-        let mut result = Ok(Vec::new());
+        let mut result = Ok(rusk_abi::RawResult::empty());
 
         if let Some((contract_id, fn_name, fn_args)) = tx.call {
-            result = rusk_abi::call_raw(
-                ContractId::from_bytes(contract_id),
-                &fn_name,
-                &fn_args,
+            let contract_id = ContractId::from_bytes(contract_id);
+            self.gas_price = Some(tx.fee.gas_price);
+            result = rusk_abi::call_raw(contract_id, &fn_name, &fn_args);
+            self.gas_price = None;
+            if let Ok(RawResult {
+                data: _,
+                economic_mode,
+            }) = result.clone()
+            {
+                match economic_mode {
+                    EconomicMode::Charge(charge) if charge != 0 => {
+                        rusk_abi::set_charge(charge)
+                    }
+                    EconomicMode::Allowance(allowance) if allowance != 0 => {
+                        rusk_abi::set_allowance(allowance)
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        result.map(|r| r.data)
+    }
+
+    /// Applies contract's charge. Caller of the contract will pay a
+    /// larger fee so that contract can earn the difference between
+    /// charge and the actual cost of the call.
+    /// Charge has no effect if the actual cost of the call is greater
+    /// than the charge.
+    /// Returns economic gas spent:
+    /// - charge if it has been applied
+    /// - gas spent if the charge has not been applied.
+    fn apply_charge(
+        &mut self,
+        contract_id: &ContractId,
+        charge: u64,
+        ges_spent: u64,
+        gas_price: u64,
+    ) -> u64 {
+        let cost = ges_spent * gas_price;
+        if charge > cost {
+            let earning = charge - cost;
+            self.add_balance(*contract_id, earning);
+            rusk_abi::emit(
+                "earning",
+                EconomicEvent {
+                    module: contract_id.to_bytes(),
+                    value: earning,
+                    result: EconomicResult::ChargeApplied,
+                },
             );
+            charge
+        } else {
+            rusk_abi::emit(
+                "earning",
+                EconomicEvent {
+                    module: contract_id.to_bytes(),
+                    value: charge,
+                    result: EconomicResult::ChargeNotSufficient,
+                },
+            );
+            ges_spent
         }
+    }
 
-        result
+    /// Applies contract's allowance. Caller of the contract's method
+    /// won't pay a fee and all the cost will be covered by the contract.
+    /// Allowance has no effect if contract does not have enough funds or
+    /// if the actual cost of the call is greater than allowance.
+    /// Returns economic gas spent:
+    /// - 0 if allowance has been applied
+    /// - gas spent if allowance has not been applied.
```

Comment:
Returning 0 if allowance has been applied mean we're losing information.
Wouldn't an enum fit better?

Reply:
we are not losing information as allowance given is equal the gas_spent - we only lose the upper limit

Reply:
this is a private subroutine